File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int size = s.size();
        if(size == 0) return 0;
        int res = 0;
        int left =0, right = 0;
        unordered_map<char, int> map;
        for(;right<size;right++){
            char c = s[right];
            map[c]++;
            while(map[c] > 1){
                map[s[left]]--;
                left++;
            }
            res = max(res, right-left + 1);
        }
        
        return res;
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int count = 0;
        int max = 0;
        int start = 0;
        Map<Character, Integer> mp = new HashMap<>();
        for(int i=0;i<s.length();i++){
            if(mp.getOrDefault(s.charAt(i), 0) > 0){
                while(start < s.length() && mp.get(s.charAt(i)) > 0){
                    mp.computeIfPresent(s.charAt(start), (key, value) -> value-1); 
                    count--;
                    start++;
                }
            }
            mp.merge(s.charAt(i), 1, (oldValue, newValue) -> oldValue + newValue);
            count++;
            max = Math.max(max, count);
        }
        return max;
    }
}